pulsar-client-api/src/main/java/org/apache/pulsar/client/api/TypedMessageBuilder.java

```diff
@@ -103,6 +103,15 @@
      */
     TypedMessageBuilder<T> keyBytes(byte[] key);
 
+    /**
+     * Sets the ordering key of the message for message dispatch in {@link SubscriptionType#Key_Shared} mode.
+     * Partition key Will be used if ordering key not specified
+     *
+     * @param orderingKey the ordering key for the message
+     * @return the message builder instance
+     */
+    TypedMessageBuilder<T> orderingKey(byte[] orderingKey);
```

Comment:
This is hugely confusing. Now we can have 2 different keys on the message. Users might think the other key doesn't guarantee ordering.

Reply:
We have already discussed through the discuss email of PIP-34,  `orderingKey` is not necessary, if user use `partitionKey` for different purposes,  `orderingKey` can offer a way allow user to ordering message by another identification.

Reply:
That's very confusing to have 2 keys

Reply:
Ok, do you have any suggestions? We really need another flag to overwrite the `partitionKey` based ordering.

Reply:
IMO It is commonly seen in a distributed database having two kind of keys, partition key and ordering key (or local key). E.g. Spanner. CDC for such databases is one of the use cases for this. Having the ability to provide an ordering key is necessary.

If it is the java doc making things confusing, we can consider improving the java doc.

Reply:
What would be the use case for having 2 keys? 

If you have partitions and set the partition routing based on 1 key, you will have ordering based on that key. If on the dispatcher you set a different key that won't address that you'll get messages out-of-order for the same orderingKey. 

My main concern is that "orderingKey" is really a dispatch key, while the ordering is guaranteed by the other key.

Reply:
> What would be the use case for having 2 keys?

@merlimat  I have explained one of the use cases above - CDC for a distributed database which usually has 2 kinds of keys. there might be other use cases.

> If you have partitions and set the partition routing based on 1 key, you will have ordering based on that key. If on the dispatcher you set a different key that won't address that you'll get messages out-of-order for the same orderingKey.

first of all, from use case perspective, I think how applications define a partition key and an ordering key is really up to themselves. but having a way to specify an ordering key provides a flexible mechanism for people to adopt to their own use cases.

secondly, from the motivation of introducing key_shared, it is allowing scaling out the consumption on a single partition while maintaining some sort of key ordering. The original "key" does already determines which partition to store message. The key indicates the ordering of the messages within a same partition, and a same "key". However that means in dispatching the messages, all the messages of a same key can only be sent to one consumer. This can limit the capability of key_shared subscription. the ordering key allows applications to configure a finer granularity "ordering_key" for scaling out the consumption beyond the original "key".

for example, in a social app, you have a conversation stream, where the stream is partitioned by `from_user_id`, so that all the conversations of a same "from_user" is in one partition, but you want to consumer to consume events by `conversation_id` (which is comprised of from_user_id and to_user_id). in this example, you are using "from_user_id" as the key, "conversation" as the ordering key.

Reply:
> @merlimat I have explained one of the use cases above - CDC for a distributed database which usually has 2 kinds of keys. there might be other use cases.

I think that's a bit generic at this point. I don't think it's even possible to do CDC on Spanner. In any case I don't see why would that be strictly required for this feature.

As always, I think it's better to add things in the API when there is a concrete need, rather than speculate possible use cases that might not apply.

> first of all, from use case perspective, I think how applications define a partition key and an ordering key is really up to themselves. but having a way to specify an ordering key provides a flexible mechanism for people to adopt to their own use cases.

The "partition-key" is what define the ordering guarantee. If you define another key it's not the "ordering-key", at most it would a sub-key, but the ordering in the log is defined by the partition key.


> However that means in dispatching the messages, all the messages of a same key can only be sent to one consumer. This can limit the capability of key_shared subscription. the ordering key allows applications to configure a finer granularity "ordering_key" for scaling out the consumption beyond the original "key".

> for example, in a social app, you have a conversation stream, where the stream is partitioned by from_user_id, so that all the conversations of a same "from_user" is in one partition, but you want the consumers to consume events by conversation_id (which is comprised of from_user_id and to_user_id). in this example, you are using "from_user_id" as the key, "conversation" as the ordering key.

In this case, since the application expect messages in order by `conversation_id`, using that as the partitioning key will achieve the same identical behavior.

>  where the stream is partitioned by from_user_id, so that all the conversations of a same "from_user" is in one partition

Why would you care about routing per `user_id` if you just care of ordering per `partition_id`?

Finally, as mentioned above I think that "ordering-key" is a very misleading name. It really would be a "sub-key", "delivery-key", "dispatch-key" or other name.

Reply:
> I think that's a bit generic at this point. I don't think it's even possible to do CDC on Spanner. In any case I don't see why would that be strictly required for this feature.

I just used Spanner as an example here. Whether Spanner supports CDC is not the point to discuss. There are many open source Spanner-like NewSQL databases. E.g. TiDB, YugaByte, and many in-house solutions. I know there are already people working on integrations between TiDB and Pulsar, where the ordering key shines there.

> As always, I think it's better to add things in the API when there is a concrete need, rather than speculate possible use cases that might not apply.

Why do you think there is no concrete need when people propose a new PIP?


> In this case, since the application expect messages in order by conversation_id, using that as the partitioning key will achieve the same identical behavior.

Pulsar is a multiple subscription system. One subscription can use failover subscription, while the other subscription can use key_shared subscription. You can't force the application to choose conversation id as the partition key. As I said, how applications can use these two keys varies from their needs.

> Why would you care about routing per user_id if you just care of ordering per partition_id?

Because there are subscriptions required to consume all the events from a particular user_id. 

> Finally, as mentioned above I think that "ordering-key" is a very misleading name. It really would be a "sub-key", "delivery-key", "dispatch-key" or other name.

I agree that "ordering" can have a different meaning in different context. It can mean - publish-order, log-order, consumer-order, dispatch-order, key-order. However I don't think "sub-key", "delivery-key" or "dispatch-key" is a better name than "ordering key". In some cases, the ordering key is a "sub-key", but it can be a completely different key while in other cases. Same applies to "delivery-key" or "dispatch-key".

IMO "ordering key" is not a bad name. It is a name that people already have some general ideas about it. Also people generally understand what partitions key and ordering key means. Applications can choose how to use them to adopt to their use cases.

However, I am also not particularly strong on the name itself. We could have called it others if there was a better name came up in the PIP discussion email thread.

Reply:
> Why do you think there is no concrete need when people propose a new PIP?

One thing is the per-key delivery, one other is a CDC pipeline. The 2 don't necessarely have to go together. This is for per-key delivery, then you bring up CDC. 

> IMO "ordering key" is not a bad name. It is a name that people already have some general ideas about it. Also people generally understand what partitions key and ordering key means

I think it's a very bad name in this context because it's a property on a message that implies something that is misleading. The implementation of this feature also has nothing to do with ordering, rather is give me messages with same keys. Ordering is a derivative property.

> In some cases, the ordering key is a "sub-key", but it can be a completely different key while in other cases.

In my view it's a "sub-key" because the routing is done on 2 levels, to partition and to consumers. If you have messages with same "ordering-key" they will be delivered out of order if they have a different partition key.

> We could have called it others if there was a better name came up in the PIP discussion email thread.

Just to understand, where is it written that things are set in stone? If people are busy to comments for a few days, one just submits a PR, his buddy approves and merge and that's it? Done? No one can comment on it anymore?

Also, as a curtesy, it would be nice to actively seek comments from people when introducing major features. It's not a race to get PR merged while others are not looking.

Reply:
> One thing is the per-key delivery, one other is a CDC pipeline. The 2 don't necessarely have to go together. This is for per-key delivery, then you bring up CDC.

CDC pipeline is a most typical use case for per-key delivery. The bottleneck of a CDC pipeline is key ordering. You want to ensure key ordering but you also want to scale out. That's why you need a per-key delivery.  I am not sure why do you think they are not related.

>  The implementation of this feature also has nothing to do with ordering, rather is give me messages with same keys. Ordering is a derivative property.

The change includes two parts: 1) adding an ordering key to allow people choose a different key for defining its ordering rather than partition key. That is the change for adding "ordering key". Hence the ordering key is for messages, not for subscription. 2) key_shared subscription is one type of subscription that choose which key to be used for dispatching messages. There can many other different ways for using these two keys.

> In my view it's a "sub-key" because the routing is done on 2 levels, to partition and to consumers. 

when you are using a term of "sub-key", it means you need both partition-key and sub-key together for deciding the ordering. in my example explained above, "from_user_id" and "conversation_id" might have some application specific relationship. but "conversation_id" is not necessarily a "sub-key" of "from_user_id". Instead, "conversation_id" provides an alternative way for grouping and ordering conversations than using "from_user_id". 

To me it is more of a name. I am fine with whether it should be called "sub-key" or "ordering-key". To me it is more a matter for making a good clarification on how `ordering_key` is used on javadoc and website.

> Just to understand, where is it written that things are set in stone? If people are busy to comments for a few days, one just submits a PR, his buddy approves and merge and that's it? Done? No one can comment on it anymore?
> Also, as a curtesy, it would be nice to actively seek comments from people when introducing major features. It's not a race to get PR merged while others are not looking.

You are overreacting to what I said here. I just meant the name was coming up from the discussion thread. I also didn't say it is a final. It can be any other names if there is a better one. In the whole conversation here, what I was trying to do is to show you the use cases I learned and my opinions - why I think "sub-key", "delivery-key" and "dispatch-key" is not better than "ordering-key".

Reply:
> when you are using a term of "sub-key", it means you need both partition-key and sub-key together for deciding the ordering. in my example explained above, "from_user_id" and "conversation_id" might have some application specific relationship. but "conversation_id" is not necessarily a "sub-key" of "from_user_id". Instead, "conversation_id" provides an alternative way for grouping and ordering conversations than using "from_user_id".

On a partition topic the ordering will be on 2 layers. Ordering will be based on the combination of both. 

If you say these 2 keys are not related, then, on a partitioned topic, the "ordering-key" will not be able to guarantee the ordering, which sounds a bit weird...

I'm not saying to use any of the examples which just came out of the mind. I'm saying that "orderingKey" can be very confusing to users because it seems to be implying something which is not.

What the application means with the key is up to the application. What it means on the dispatcher is that, *within a given partition*, all messages with same key will go to same consumer. "Ordering-Key" as a name fails to convey any of the properties while clashing with the partition key that is already used to decide the ordering on the partitions.

To throw more names:  "consumeOrderingKey", "dispatchOrderingKey" 

> You are overreacting to what I said here. I just meant the name was coming up from the discussion thread. I also didn't say it is a final. 

Am I? Perhaps it was not intended but your phrase sounded exactly like that.

Reply:
frankly speaking, I am not sure it is a good idea to add "consume" or "dispatch" to the "ordering-key". 

in a data system, "partition-key" and "ordering-key" is actually a clean concept. "partition-key" is used for partitioning and "ordering-key" is used for ordering. ideally a partition-key doesn't have any ordering implication. but because pulsar is implementing a partition based on an ordered log, the "ordering" on a "partition-key" is implicitly indicated.

If an application needs end-to-end ordering, it can omit "ordering-key" or choose a sub-key of "partition-key" or another related key. since end-to-end ordering requires producers and consumers have agreements on keys.

If an application doesn't require end-to-end ordering or only needs partial ordering (e.g. in a map-reduce-ish pipeline, a topic is used for computing partially aggregated state), it can just use a key for partitioning key and the other key for ordering key.

decoupling the partition-key and ordering-key allows adopting to more use case in future.

However as I said in previous comment, I don't have any strong opinions about the naming. I would let the original authors and you decide which name to pick.